ssl/ssl_lib.c

```diff
@@ -702,27 +702,26 @@ SSL *SSL_new(SSL_CTX *ctx)
 
 int ossl_ssl_init(SSL *ssl, SSL_CTX *ctx, const SSL_METHOD *method, int type)
 {
-    ssl->type = type;
-
     ssl->lock = CRYPTO_THREAD_lock_new();
     if (ssl->lock == NULL)
         return 0;
 
     if (!CRYPTO_NEW_REF(&ssl->references, 1)) {
         CRYPTO_THREAD_lock_free(ssl->lock);
+        ssl->lock = NULL;
         return 0;
     }
 
-    if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_SSL, ssl, &ssl->ex_data)) {
+    if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_SSL, ssl, &ssl->ex_data)
+            || !SSL_CTX_up_ref(ctx)) {
```

Comment:
We should up ref first and do everything else afterwards, freeing the SSL_CTX if anything further fails.